File: 1855-maximum-distance-between-a-pair-of-values/1855-maximum-distance-between-a-pair-of-values.rb
module Comparable
    def max(v) = (self > v ? self : v)
end

def max_distance(nums1, nums2)
    return 0 if nums1.last > nums2.first
    n1, n2 = nums1.size, nums2.size
    i1 = 0
    nums2.each_with_index.inject(0) {|res, (v2, i2)|
        i1 += 1 while i1 < i2 && i1 < n1 && nums1[i1] > v2
        return res if i1 >= n1
        res.max(i2 - i1)
    }
end

File: 1855-maximum-distance-between-a-pair-of-values/1855-maximum-distance-between-a-pair-of-values.cpp
class Solution {
public:
    int maxDistance(vector<int>& nums1, vector<int>& nums2) {
        int n=nums1.size();
        reverse(nums2.begin(),nums2.end());
        int m=nums2.size();
        int maxi=0;
        for(int i=0;i<n;i++){
            auto idx=std::lower_bound(nums2.begin(),nums2.end(),nums1[i])-nums2.begin();
          // cout<<i<<idx<<endl;
               // idx=idx-nums2.begin();
             // int j=idx-nums2.begin();
             if(idx==nums2.size())continue;
                   int act_idx=m-1-idx;
                    maxi=max(maxi,act_idx-i);
                   
                }
            
            return maxi;
            

        }
};